### PR TITLE
py-horovod: set appropriate build env vars

### LINKED ISSUE #15545: Installation issue: py-horovod
The py-horovod package doesn't seem to actually install anything.  Maybe @rvinaybharadwaj or @adamjstewart have an idea on how to fix it?

### Spack version
```console
$ spack --version
0.14.0-247-958f260
```

### Steps to reproduce the issue

```console
$ spack install py-horovod
$ find ~/workspace/spack/opt/spack/linux-rhel7-haswell/gcc-7.3.0/py-horovod-0.19.0-6bfj56gcf2pegrhj333s3glwks6zvtdx | grep -v \\.spack
~/workspace/spack/opt/spack/linux-rhel7-haswell/gcc-7.3.0/py-horovod-0.19.0-6bfj56gcf2pegrhj333s3glwks6zvtdx
~/workspace/spack/opt/spack/linux-rhel7-haswell/gcc-7.3.0/py-horovod-0.19.0-6bfj56gcf2pegrhj333s3glwks6zvtdx/lib
~/workspace/spack/opt/spack/linux-rhel7-haswell/gcc-7.3.0/py-horovod-0.19.0-6bfj56gcf2pegrhj333s3glwks6zvtdx/lib/perl5
~/workspace/spack/opt/spack/linux-rhel7-haswell/gcc-7.3.0/py-horovod-0.19.0-6bfj56gcf2pegrhj333s3glwks6zvtdx/lib/python3.7
~/workspace/spack/opt/spack/linux-rhel7-haswell/gcc-7.3.0/py-horovod-0.19.0-6bfj56gcf2pegrhj333s3glwks6zvtdx/lib/python3.7/site-packages
```

### Platform and user environment

```commandline
$ uname -a 
Linux --- 3.10.0-1062.9.1.1chaos.ch6.x86_64 #1 SMP Mon Dec 16 13:50:10 PST 2019 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:	Red Hat Enterprise Linux Server release 7.7 (Maipo)
``` 

### Additional information

```console
$ spack spec --install-status py-horovod
Input spec
--------------------------------
 -   py-horovod

Concretized
--------------------------------
[+]  py-horovod@0.19.0%gcc@7.3.0+cuda~gloo+mpi~mxnet+pytorch arch=linux-rhel7-haswell
[+]      ^cuda@10.1.168%gcc@7.3.0 arch=linux-rhel7-haswell
[+]      ^mvapich2@2.3%gcc@7.3.0~alloca ch3_rank_bits=32 ~cuda~debug fabrics=mrail file_systems=auto process_managers=auto +regcache threads=multiple +wrapperrpath arch=linux-rhel7-haswell
[+]      ^py-cloudpickle@1.2.1%gcc@7.3.0 arch=linux-rhel7-haswell
[+]          ^py-setuptools@41.4.0%gcc@7.3.0 arch=linux-rhel7-haswell
[+]              ^python@3.7.6%gcc@7.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4~uuid+zlib arch=linux-rhel7-haswell
[+]                  ^bzip2@1.0.8%gcc@7.3.0+shared arch=linux-rhel7-haswell
[+]                      ^diffutils@3.7%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                          ^libiconv@1.16%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                  ^expat@2.2.9%gcc@7.3.0+libbsd arch=linux-rhel7-haswell
[+]                      ^libbsd@0.10.0%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                  ^gdbm@1.18.1%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                      ^readline@8.0%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                          ^ncurses@6.2%gcc@7.3.0~symlinks+termlib arch=linux-rhel7-haswell
[+]                              ^pkgconf@1.6.3%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                  ^gettext@0.20.1%gcc@7.3.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-rhel7-haswell
[+]                      ^libxml2@2.9.9%gcc@7.3.0~python arch=linux-rhel7-haswell
[+]                          ^xz@5.2.4%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                          ^zlib@1.2.11%gcc@7.3.0+optimize+pic+shared arch=linux-rhel7-haswell
[+]                      ^tar@1.32%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                  ^libffi@3.2.1%gcc@7.3.0 arch=linux-rhel7-haswell
[+]                  ^openssl@1.1.1d%gcc@7.3.0+systemcerts arch=linux-rhel7-haswell
[+]                      ^perl@5.30.1%gcc@7.3.0+cpanm+shared+threads arch=linux-rhel7-haswell
[+]                  ^sqlite@3.30.1%gcc@7.3.0~column_metadata+fts~functions~rtree arch=linux-rhel7-haswell
[+]      ^py-pip@19.3%gcc@7.3.0 arch=linux-rhel7-haswell
[+]      ^py-psutil@5.6.3%gcc@7.3.0 arch=linux-rhel7-haswell
[+]      ^py-pyyaml@5.1.2%gcc@7.3.0 arch=linux-rhel7-haswell
[+]          ^libyaml@0.2.2%gcc@7.3.0 arch=linux-rhel7-haswell
[+]      ^py-six@1.12.0%gcc@7.3.0 arch=linux-rhel7-haswell
[+]      ^py-torch@1.4.0%gcc@7.3.0~binary~caffe2+cuda cuda_arch=37 +cudnn+distributed~fbgemm~ffmpeg~gloo~leveldb~lmdb~magma~miopen+mkldnn+nccl~nnpack~opencv+openmp~qnnpack~redis~tbb~test~zstd arch=linux-rhel7-haswell
[+]          ^cmake@3.12.1%gcc@7.3.0~doc~ncurses~openssl+ownlibs patches=93488edf3f89cef6f3c77e821007c022e88e9f1f566ae9bf824a87ecbac97ebb,dd3a40d4d92f6b2158b87d6fb354c277947c776424aa03f6dc8096cf3135f5d0 ~qt arch=linux-rhel7-haswell
[+]          ^cudnn@7.6.5.32-10.1-linux-x64%gcc@7.3.0 arch=linux-rhel7-haswell
[+]          ^eigen@3.3.7%gcc@7.3.0 build_type=RelWithDebInfo arch=linux-rhel7-haswell
[+]          ^nccl@2.4.8-1%gcc@7.3.0 patches=42778c78eb9875dacddf5eca20f7f6a077773fcbee41e51174f81b3143684b6d arch=linux-rhel7-haswell
[+]          ^ninja@1.10.0%gcc@7.3.0 arch=linux-rhel7-haswell
[+]          ^openblas@0.3.6%gcc@7.3.0~consistent_fpcsr~ilp64+pic+shared threads=openmp arch=linux-rhel7-haswell
[+]          ^protobuf@3.7.1%gcc@7.3.0 build_type=Release +shared arch=linux-rhel7-haswell
[+]          ^py-numpy@1.16.2%gcc@7.3.0+blas+lapack patches=196dc7cc8e1367de199726d4e005bba50c3d13fb2843c4753c5f07f7894f10c9 arch=linux-rhel7-haswell
[+]          ^py-pybind11@2.4.3%gcc@7.3.0 build_type=RelWithDebInfo arch=linux-rhel7-haswell
```

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #15545 

@rblake-llnl @rvinaybharadwaj can you test this? 

P.S. if anyone else would like to be added as a maintainer, let me know.

### CODE REVIEW

**Comment by rblake-llnl** [comment]
With default options:

```
  >> 125    error: You should not mix NCCL and MPI GPU due to a possible deadlock.
```

When tuning everything to use mpi

```
  >> 138    horovod/common/parameter_manager.h:26:10: fatal error: Eigen/Core: No such file or directory
```

I'll dig into it more in a second.

**Comment by adamjstewart** [comment]
@rblake-llnl 

> error: You should not mix NCCL and MPI GPU due to a possible deadlock.

Fixed by setting `HOROVOD_ALLOW_MIXED_GPU_IMPL` and switching defaults.

> fatal error: Eigen/Core: No such file or directory

Fixed by checking out git submodules instead of downloading tarball

@rvinaybharadwaj 

> ValueError: could not convert string to float: 'none'

Fixed by modifying conflict statements. If you use CUDA with PyTorch or TF, you need to also specify the `cuda_arch` of your GPU.

**Comment by rvinaybharadwaj** [comment]
> @rblake-llnl
> 
> > error: You should not mix NCCL and MPI GPU due to a possible deadlock.
> 
> Fixed by setting `HOROVOD_ALLOW_MIXED_GPU_IMPL` and switching defaults.
> 
> > fatal error: Eigen/Core: No such file or directory
> 
> Fixed by checking out git submodules instead of downloading tarball
> 
> @rvinaybharadwaj
> 
> > ValueError: could not convert string to float: 'none'
> 
> Fixed by modifying conflict statements. If you use CUDA with PyTorch or TF, you need to also specify the `cuda_arch` of your GPU.

May be disable CUDA build by default

**Comment by adamjstewart** [comment]
Deep learning without GPU support isn't really useful, I think the default should be `+cuda`.

**Comment by rvinaybharadwaj** [comment]
> Deep learning without GPU support isn't really useful, I think the default should be `+cuda`.

Agreed. However, none of the upcoming exascale machines employ NVIDIA hardware.

**Comment by rblake-llnl** [comment]
Works for me, and also runs a hello-world example training mnist

**Comment by rblake-llnl** [comment]
> > Deep learning without GPU support isn't really useful, I think the default should be `+cuda`.
> 
> Agreed. However, none of the upcoming exascale machines employ NVIDIA hardware.

I also vote for `+cuda`.  GPUs are so much faster than CPUs that there isn't much point in training without support for them.  People who want to train on exascale should require to take care and take note that they're using the slow version of training.

**Comment by rvinaybharadwaj** [comment]
> > > Deep learning without GPU support isn't really useful, I think the default should be `+cuda`.
> > 
> > 
> > Agreed. However, none of the upcoming exascale machines employ NVIDIA hardware.
> 
> I also vote for `+cuda`. GPUs are so much faster than CPUs that there isn't much point in training without support for them. People who want to train on exascale should require to take care and take note that they're using the slow version of training.

My point is different machines use different GPUs and accelerators, and the install should not error out with default options.

**Comment by adamjstewart** [comment]
It sounds to me like users of new exascale machines won't even be installing Horovod/TF/PyTorch if that's the case. It's a shame that the people making purchasing decisions aren't aware of the trend towards deep learning in science, but there isn't much we can do about that.

The default variants don't error out, they raise a conflict with a message that explicitly states what they need to do, so there's no confusion. But again, installing a CPU-only version of a deep learning library will result in an order of magnitude performance drop: https://medium.com/@shachishah.ce/do-we-really-need-gpu-for-deep-learning-47042c02efe2. I don't mind the default variants requiring intervention for some users, but I do mind Spack silently installing a version of a deep learning library that's completely useless for 99% of deep learning researchers.

P.S. I'm a deep learning researcher 😄

**Comment by rvinaybharadwaj** [comment]
> It sounds to me like users of new exascale machines won't even be installing Horovod/TF/PyTorch if that's the case. It's a shame that the people making purchasing decisions aren't aware of the trend towards deep learning in science, but there isn't much we can do about that.
> 
> The default variants don't error out, they raise a conflict with a message that explicitly states what they need to do, so there's no confusion. But again, installing a CPU-only version of a deep learning library will result in an order of magnitude performance drop: https://medium.com/@shachishah.ce/do-we-really-need-gpu-for-deep-learning-47042c02efe2. I don't mind the default variants requiring intervention for some users, but I do mind Spack silently installing a version of a deep learning library that's completely useless for 99% of deep learning researchers.
> 
> P.S. I'm a deep learning researcher 😄

I don't have any objections if you feel so strongly about it. But the vendors will be supporting different frameworks on their platforms.

File: var/spack/repos/builtin/packages/py-torch/package.py [hunk 1]
```diff
@@ -105,22 +105,22 @@ class PyTorch(PythonPackage, CudaPackage):
     cuda_arch_conflict = ('This version of Torch/Caffe2 only supports compute '
                           'capabilities ')
 
-    conflicts('cuda_arch=none', when='+cuda+caffe2',
+    conflicts('cuda_arch=none', when='+cuda',
               msg='Must specify CUDA compute capabilities of your GPU, see '
               'https://developer.nvidia.com/cuda-gpus')
-    conflicts('cuda_arch=52', when='@1.3.0:+cuda+caffe2',
+    conflicts('cuda_arch=52', when='@1.3.0:+cuda',
```
Comment (darthsuogles) [inline var/spack/repos/builtin/packages/py-torch/package.py]: What's the reason behind deprecating these versions? 
At least looks like `TORCH_CUDA_ARCH_LIST="3.5 5.2 6.0 6.1 7.0+PTX" ` is still in the official docker image.
https://github.com/pytorch/pytorch/blob/c8f665dcb626378ca9c51d7b70fc242524105441/Dockerfile#L47

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-torch/package.py]: These conflicts were added by @glennpj in #14619, maybe he can explain where they came from. Hopefully I didn't misunderstand the code and the restriction only applies to `+cuda+caffe2`.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-torch/package.py]: Ping @glennpj. I was able to build PyTorch 1.4.0 with `cuda_arch=35` just fine (although I haven't tested it yet). Where did these conflicts come from?

**Pull request merged**